cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+      gen_list.push_back(col_idx);
+    }
+    return schema(std::move(fields));
+  }
+}
+
+Result<Datum> RandomVString(random::pcg32_fast& rng, int64_t num_rows, int32_t min_length,
+                            int32_t max_length) {
+  std::uniform_int_distribution<int32_t> length_dist(min_length, max_length);
+  ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> offset_buff,
+                        AllocateBuffer((num_rows + 1) * sizeof(int32_t)));
+  int32_t* offsets = reinterpret_cast<int32_t*>(offset_buff->mutable_data());
+  offsets[0] = 0;
+  for (int64_t i = 1; i <= num_rows; i++) offsets[i] = offsets[i - 1] + length_dist(rng);
+  ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> str_buff,
+                        AllocateBuffer(offsets[num_rows]));
+  char* str = reinterpret_cast<char*>(str_buff->mutable_data());
+
+  // Spec says to pick random alphanumeric characters from a set of at least
+  // 64 symbols. Now, let's think critically here: 26 letters in the alphabet,
+  // so 52 total for upper and lower case, and 10 possible digits gives 62
+  // characters...
+  // dbgen solves this by including a space and a comma as well, so we'll
+  // copy that.
+  const char alpha_numerics[65] =
+      "0123456789abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLMNOPQRSTUVWXYZ,";
+  std::uniform_int_distribution<int> char_dist(0, 63);
+  for (int32_t i = 0; i < offsets[num_rows]; i++) str[i] = alpha_numerics[char_dist(rng)];
+
+  ArrayData ad(utf8(), num_rows, {nullptr, std::move(offset_buff), std::move(str_buff)});
+  return std::move(ad);
+}
+
+void AppendNumber(char*& out, int num_digits, int32_t x) {
+  out += (num_digits - 1);
+  while (x > 0) {
+    *out-- = '0' + (x % 10);
+    x /= 10;
+  }
+  out += (num_digits + 1);
+}
+
+void GeneratePhoneNumber(char* out, random::pcg32_fast& rng, int32_t country) {
+  std::uniform_int_distribution<int32_t> three_digit(100, 999);
+  std::uniform_int_distribution<int32_t> four_digit(1000, 9999);
+
+  int32_t country_code = country + 10;
+  int32_t l1 = three_digit(rng);
+  int32_t l2 = three_digit(rng);
+  int32_t l3 = four_digit(rng);
+  AppendNumber(out, 2, country_code);
+  *out++ = '-';
+  AppendNumber(out, 3, l1);
+  *out++ = '-';
+  AppendNumber(out, 3, l2);
+  *out++ = '-';
+  AppendNumber(out, 4, l3);
+}
+
+static constexpr uint32_t kStartDate =
+    8035;  // January 1, 1992 is 8035 days after January 1, 1970
+static constexpr uint32_t kCurrentDate =
+    9298;  // June 17, 1995 is 9298 days after January 1, 1970
+static constexpr uint32_t kEndDate =
+    10591;  // December 12, 1998 is 10591 days after January 1, 1970
+
+using GenerateColumnFn = std::function<Status(size_t)>;
+class PartAndPartSupplierGenerator {
+ public:
+  Status Init(size_t num_threads, int64_t batch_size, float scale_factor) {
+    if (!inited_) {
+      inited_ = true;
+      batch_size_ = batch_size;
+      scale_factor_ = scale_factor;
+
+      arrow_vendored::pcg_extras::seed_seq_from<std::random_device> seq;
```

Comment:
It would be better to call `GetRandomSeed` (see `arrow/util/io_util.h`).

Reply:
Also, I'm curious: is this deliberately non-deterministic between runs? @westonpace

Reply:
i switched it to use GetRandomSeed. in regards to nondeterminism between runs i'll let Weston decide.

Reply:
As an end user of this code, I would appreciate the ability to provide determinism if that's not too painful.

We wouldn't always want or need it, but sometimes it is helpful to be able to generate the exact same set of data.

Reply:
Agreed.  We can take a seed in the constructor.  If not specified I think defaulting to non-deterministic is ok.

Reply:
i made it take a seed, defaulting to generating a nondeterministic one.